### DISCUSSION TITLE
How to implement unit testing in directus

### DISCUSSION THREAD

**Discussion by moayadsquad** [directus/directus#12191] [Archive]
How to implement unit testing that test API,
My main database is: MySQL , I want to test with sqlite3 (in memory), that will not affect my main database, also how to make migration or apply snapshot to sqlite3?

**Comment by rijkvanzanten** [comment]
You wanna test the Directus API? In that case, I'd recommend just adding on to the tests we already started: https://github.com/directus/directus/tree/main/tests/api

**Reply by rijkvanzanten** [reply]
Also keep in mind that you might not want to test on SQLite3 if you use MySQL in production. There could be unexpected (subtle) differences or bugs that only happen in one or the other

**Reply by moayadsquad** [reply]
Thank you @rijkvanzanten, but I need to test my custom API + custom hook validations, I don't want to make test on my actual database, I want with SQLite3 (in memory database), how to  apply snapshot from MySQL to SQLite3? is this applicable on directus?

**Reply by rijkvanzanten** [reply]
There's no directus utility available to apply an existing mysql database on top of sqlite

**Reply by moayadsquad** [reply]
is there any third party library that help me to do that? or there any work around to apply unit testing on an existing MySQL with SQLite3?

**Reply by rijkvanzanten** [reply]
No idea! You can't just dump one SQL database onto another, as they often have differences in data types / identifiers / quotation styles etc etc.
